### PR TITLE
Adds bitset

### PR DESCRIPTION
**Pull request by FloweyTheFlower420**
Adds a somewhat c++ standard conforming bitset. Some features have not been tested, and the formatting is probably incorrect.

### CODE REVIEW

File: include/frg/bitset.hpp [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        constexpr bool operator[](size_t pos) const { return get_bit(buffer[pos / 64], pos % 64); }
+
+        reference operator[](size_t pos) { return reference(pos, *this); }
+
+        size_t count() const noexcept
+        {
+            size_t n = 0;
+            for (size_t i = 0; i < buffer_size - 1; i++)
+                n += __builtin_popcountll(buffer[i]);
+            return n + __builtin_popcountll(get_bits(buffer[buffer_size - 1], 0, N % 64));
```
Comment (ArsenArsen) [inline include/frg/bitset.hpp:178]: This probably isn't needed (the upper bits of this bufer element should always be zero)

File: include/frg/bitset.hpp [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+            if (N % 64)
+                value &= (buffer[N / 64] & ((1 << (N % 64)) - 1)) == (1 << (N % 64)) - 1;
+            return value;
+        }
+
+        bool any() const noexcept
+        {
+            bool value = true;
+            for (size_t i = 0; i < N / 64; i++)
+                value |= (buffer[i] == 0xffffffffffffffff);
```
Comment (ArsenArsen) [inline include/frg/bitset.hpp:206]: isn't this an all algorithm?

**Review by ArsenArsen — CHANGES_REQUESTED** [review]
also, please reword the commit to align with the rest of the repo and add unit tests

thanks!

**Review by qookei — CHANGES_REQUESTED** [review]
The code style also doesn't match the one we use here. The code should be indented with tabs, opening brackets go on the same line, pointers and references are right-side (`T &a` not `T& a`). Commit message also doesn't look right. The code itself looks good to me though.

**Comment by FloweyTheFlower420** [comment]
Do you have a .clang-format or something similar for formatting?

**Comment by no92** [comment]
We have a [style guide](https://docs.managarm.org/handbook/contributing/coding-style.html).

File: include/frg/bitset.hpp [hunk 3]
```diff
@@ -0,0 +1,248 @@
+#ifndef FRG_BITSET_HPP
+#define FRG_BITSET_HPP
+#include <cstddef>
+#include <cstdint>
+
+namespace frg {
+	namespace detail {
+		constexpr auto div_roundup(auto v1, auto v2) { return (v1 + v2 - 1) / v2; }
```
Comment (no92) [inline include/frg/bitset.hpp:8]: Given that the two `auto` parameters basically turn this into a templated function, the signedness of those params is left to chance - which might result in undefined behavior (signed int overflow is undefined, while unsigned ints wrap around).

File: tests/meson.build [hunk 4]
```diff
@@ -3,9 +3,9 @@ gtest = dependency('gtest_main')
 test_executable = executable('frigg_tests',
 	'tests.cpp',
 	dependencies: [
-		gtest, frigg_dep
+		gtest, frigg_dep 
 	],
-	override_options: ['cpp_std=c++20'],
```
Comment (no92) [inline tests/meson.build:8]: Changing the warning level should be a separate commit, IMO.

**Review by no92 — COMMENTED** [review]
With the styling commit rolled into the commits changing the code, and with the comments addressed, LGTM

File: include/frg/bitset.hpp (hunk 3 above)
Comment (ArsenArsen) [reply include/frg/bitset.hpp:8]: Likely best to use size_t (and maybe handle the overflows)

**Comment by ArsenArsen** [comment]
We don't indent namespaces, and the tests aren't indented properly still

**Review by no92 — APPROVED** [review]
If rebased and the last three commits reworked into the first two commits, LGTM

**Comment by FloweyTheFlower420** [comment]
Could y'all see if this can be merged?

File: tests/tests.cpp [hunk 5]
```diff
@@ -196,8 +191,8 @@ TEST(formatting, fmt) {
 	EXPECT_EQ(str, "{:h}");
 	str.clear();
 
-	std::string abc_def { "abc def" };
-	std::vector<char> abc_def_v { abc_def.cbegin(), abc_def.cend() };
+	std::string abc_def{"abc def"};
+	std::vector<char> abc_def_v{abc_def.cbegin(), abc_def.cend()};
```
Comment (ArsenArsen) [inline tests/tests.cpp:195]: these changes appear to be unrelated

File: tests/meson.build [hunk 6]
```diff
@@ -3,7 +3,7 @@ gtest = dependency('gtest_main')
 test_executable = executable('frigg_tests',
 	'tests.cpp',
 	dependencies: [
-		gtest, frigg_dep
+		gtest, frigg_dep
```
Comment (ArsenArsen) [inline tests/meson.build:6]: this change appears to be unrelated (or unwanted even?)

**Review by ArsenArsen — CHANGES_REQUESTED** [review]
algorithmically, this looks okay, note that libstdc++ is making <bitset> freestanding as of gcc13 (or, as freestanding as it gets), so maybe you want to account for that in order to write code that can easily migrate to the standard bitset

the few last comments are very simple, so when those are fixed up I can merge this.

also, this assumes `CHAR_BIT` is eight, might want to add a static assert about that

**Comment by FloweyTheFlower420** [comment]
I'm not precisely sure what you mean by "writing code that can easily migrate to the standard bitset," but the other changes have been implemented.

**Comment by ArsenArsen** [comment]
what I mean is that c++ already provides a [`<bitset>`](https://en.cppreference.com/w/cpp/utility/bitset) so it might be valuable to match the interface that one provides so that one can switch between it and frigg easily
